File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        q = deque([start])
        while q:
            cur = q.popleft()
            if arr[cur] == 0:
                return True
            if arr[cur] <0: 
                continue
            if cur + arr[cur] < len(arr):
                q.append(cur + arr[cur])
            if cur - arr[cur] >= 0:
                q.append(cur-arr[cur])
            arr[cur] *= -1
            
        return False
                
            
            
        

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& A, int start) {
        queue<int> q;
        q.push(start);
        while(size(q)) {
            int cur = q.front(); q.pop();
            if(A[cur] == 0) return true;                        // target reached
            if(A[cur] < 0)  continue;                           // already visited - stop further exploration of this path
            if(cur + A[cur] < size(A)) q.push(cur + A[cur]);    // try forward jump
            if(cur - A[cur] >= 0)      q.push(cur - A[cur]);    // try backward jump
            A[cur] *= -1;                                       // mark current index as visited by negating
        }
        return false;     
    }
};